packages/utils/crypto/src/encryption.ts

```diff
@@ -1,19 +1,218 @@
+	parameters: PublicKeyEncryptionKeys | SymmetricKeys
+): parameters is PublicKeyEncryptionKeys {
+	return (parameters as PublicKeyEncryptionKeys) instanceof X25519Keypair;
+}
+
+type EnvelopeFromParameter<Parameters extends KeyExchangeOptions> =
+	Parameters extends PublicKeyEncryptionParameters
+		? PublicKeyEnvelope
+		: HashedKeyEnvelope;
+
+type EncryptProvide<Parameters extends KeyExchangeOptions> = (
+	bytes: Uint8Array,
+	parameters: Parameters
+) => Promise<CipherWithEnvelope<EnvelopeFromParameter<Parameters>>>;
+
+export const createLocalEncryptProvider = <
+	K extends PublicKeyEncryptionKeys | SymmetricKeys,
+	Parameters extends KeyExchangeOptions = K extends PublicKeyEncryptionKeys
+		? PublicKeyEncryptionParameters
+		: SymmetricKeyEncryptionParameters
+>(
+	keys: K
+) => {
+	return async (
+		bytes: Uint8Array,
+		parameters: Parameters
+	): Promise<CipherWithEnvelope<EnvelopeFromParameter<Parameters>>> => {
+		const nonce = randomBytes(NONCE_LENGTH); // crypto random is faster than sodim random
+		if (
+			isAsymmetriEncryptionParameters(parameters) &&
+			isAsymmetricEncryptionKeys(keys)
+		) {
+			const epheremalKey = sodium.crypto_secretbox_keygen();
+			const cipher = sodium.crypto_secretbox_easy(bytes, nonce, epheremalKey);
+			const { receiverPublicKeys } = parameters;
+			const receiverX25519PublicKeys = await Promise.all(
+				receiverPublicKeys.map((key) => {
+					if (key instanceof Ed25519PublicKey) {
+						return X25519PublicKey.from(key);
+					}
+					return key;
+				})
+			);
+
+			const ks = receiverX25519PublicKeys.map((receiverPublicKey) => {
+				const kNonce = randomBytes(NONCE_LENGTH); // crypto random is faster than sodium random
+				return new K({
+					encryptedKey: new CipherWithNonce({
+						cipher: sodium.crypto_box_easy(
+							epheremalKey,
+							kNonce,
+							receiverPublicKey.publicKey,
+							keys.secretKey.secretKey
+						),
+						nonce: kNonce
+					}),
+					receiverPublicKey
+				});
+			});
+
+			return {
+				cipher: new Uint8Array(cipher), // TODO do we need this clone?
+				nonce,
+				envelope: new PublicKeyEnvelope({
+					senderPublicKey: keys.publicKey,
+					ks
+				}) as EnvelopeFromParameter<Parameters>
+			};
+		} else if (
+			!isAsymmetriEncryptionParameters(parameters) &&
+			!isAsymmetricEncryptionKeys(keys)
+		) {
+			const cipher = sodium.crypto_secretbox_easy(bytes, nonce, keys);
+			return {
+				cipher: new Uint8Array(cipher), // TODO do we need this clone?
+				nonce,
+				envelope: new HashedKeyEnvelope({
+					hash: await sha256(keys)
+				}) as EnvelopeFromParameter<Parameters>
+			};
+		}
+
+		throw new Error("Unexpected encryption parameters");
+	};
+};
+
+type DecryptProvider = (
+	encrypted: Uint8Array,
+	nonce: Uint8Array,
+	exchange: Envelope
+) => Promise<Uint8Array>;
+
+type KeyResolver = <PublicKey extends X25519PublicKey | Uint8Array>(
+	key: PublicKey
+) =>
+	| (PublicKey extends X25519PublicKey ? X25519Keypair : Uint8Array)
```

Comment:
Could we return `Keypair` here instead of `Uint8Array`?

Reply:
The Uint8array referred to the symmetric key with length 32.. the one that is going to be used with XSalsa20Poly1305 encryption scheme